File: runme/mcgen/local/datan/datsrc/lsqlin.c
#include "datsrc.h"

void LIBFUNC lsqlin_(double const P_T *t, double P_T *c,
             double const P_T *deltay, integer const P_T *n, integer P_T *nr,
             double P_T *x, double P_T *cx, double P_T *r, double P_T *a,
             double P_T *scrat, logical P_T *ok)
{
    /* Table of constant values */

    integer c__1 = 1;
    double c_d0 = 0.;

    /* System generated locals */
    integer a_dim1, a_offset, cx_dim1, cx_offset, scrat_dim1, scrat_offset, 
            i__1, i__2;

    /* Local variables */
    integer i, k;

/* compute matrix A' from A */
    /* Parameter adjustments */
    --deltay;
    --c;
    --t;
    scrat_dim1 = *nr;
    scrat_offset = scrat_dim1 + 1;
    scrat -= scrat_offset;
    a_dim1 = *n;
    a_offset = a_dim1 + 1;
    a -= a_offset;
    cx_dim1 = *nr;
    cx_offset = cx_dim1 + 1;
    cx -= cx_offset;
    --x;

    /* Function Body */
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        i__2 = *nr;
        for (k = 1; k <= i__2; ++k) {
            a[i + k * a_dim1] /= deltay[i];
        }
    }
/* Set up vector C' */
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        c[i] /= deltay[i];
    }
/* Set up matrix GX */
    mtxmat_(&a[a_offset], &a[a_offset], &cx[cx_offset], nr, n, nr);
/* Determine vector X of unknowns */
    mtxsvd_(&a[a_offset], &c[1], &x[1], r, n, nr, &c__1, &c_d0, ok);
    if (*ok) {
/* Determine covariance matrix CX by inverting GX */
        mtxchi_(&cx[cx_offset], &scrat[scrat_offset], nr);
    }
    return;
} /* lsqlin_ */

File: runme/mcgen/local/datan/datsrc/lsqlin.f
      SUBROUTINE LSQLIN(T,C,DELTAY,N,NR,X,CX,R,A,SCRAT,OK)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION A(N,NR),T(N),C(N),DELTAY(N),X(NR),CX(NR,NR)
      DIMENSION SCRAT(NR,NR)
      LOGICAL OK
C compute matrix A' from A
      DO 20 I=1,N
        DO 10 K=1,NR
          A(I,K)=A(I,K)/DELTAY(I)
   10   CONTINUE
   20 CONTINUE
C Set up vector C'
      DO 30 I=1,N
        C(I)=C(I)/DELTAY(I)
   30 CONTINUE
C Set up matrix GX
      CALL MTXMAT(A,A,CX,NR,N,NR)
C Determine vector X of unknowns
      CALL MTXSVD(A,C,X,R,N,NR,1,DBLE(0.),OK)
      IF(OK) THEN
C Determine covariance matrix CX by inverting GX
        CALL MTXCHI(CX,SCRAT,NR)
      END IF
      END
